### PR TITLE
Improve logging

### PR DESCRIPTION
**Pull request by andrenpaes**
Adds latency to log entries, only logging one event per transaction (with an error flag attribute included).

Also refactors a bit the notification handling code, also adding latency to log entries.

Here are a few examples of log entries for Transaction Operations:
```bash
# Hpp
op='createHppRequest', merchantAccount='MyCityDealDE', amount='10', amount='10', paymentTransactionExternalKey='fd623240-262f-4f0c-b32f-22f5e6f25199', error=false

# Payments
op='authorize', merchantAccount='MyCityDealDE', amount='10', amount='10', paymentTransactionExternalKey='eb5859d2-3691-4b63-b504-7f7c7a820526', PurchaseResult{result='Authorised', authCode='75613', pspReference='8814725091152670', reason='null', resultCode='Authorised', reference='null', paymentTransactionExternalKey='eb5859d2-3691-4b63-b504-7f7c7a820526', adyenCallErrorStatus=null, additionalData={authCode='75613', refusalReasonRaw='AUTHORISED', avsResult='5 No AVS data provided', avsResultRaw='5', acquirerReference='7F54NDV1I6P', cvcResult='1 Matches', cvcResultRaw='1', acquirerCode='TestPmmAcquirer'}}, duration=3024, error=false
op='cancel', merchantAccount='MyCityDealDE', pspReference='1a24b4f3-a4b5-4ddc-9bd4-16dcdec207fd', paymentTransactionExternalKey='d3fae27a-eeb3-4baa-ba5b-02be55ec8ce4', pspReference='1a24b4f3-a4b5-4ddc-9bd4-16dcdec207fd', UnSuccessfulAdyenCall{responseStatus=RESPONSE_INVALID, exceptionMessage='HTTP response '422: Unprocessable Entity' when communicating with https://pal-test.adyen.com/pal/servlet/Payment/v12', exceptionClass='org.apache.cxf.transport.http.HTTPException', duration=176 }, error=true

op='authorize', merchantAccount='MyCityDealDE', amount='10', amount='10', paymentTransactionExternalKey='2485f71c-1cfa-4baf-ba24-cf09c6732954', PurchaseResult{result='Authorised', authCode='23053', pspReference='8514725091166343', reason='null', resultCode='Authorised', reference='null', paymentTransactionExternalKey='2485f71c-1cfa-4baf-ba24-cf09c6732954', adyenCallErrorStatus=null, additionalData={authCode='23053', refusalReasonRaw='AUTHORISED', avsResult='5 No AVS data provided', avsResultRaw='5', acquirerReference='7CA6B5NDF1P', cvcResult='1 Matches', cvcResultRaw='1', acquirerCode='TestPmmAcquirer'}}, duration=965, error=false
op='capture', merchantAccount='MyCityDealDE', amount='5', amount='5', paymentTransactionExternalKey='75679349-688d-4f5c-b14c-e6f9f72e0519', pspReference='8514725091166343', PaymentModificationResponse{adyenCallErrorStatus=null, pspReference='8614725091175632', response='[capture-received]', additionalData={}}, duration=181, error=false
op='capture', merchantAccount='MyCityDealDE', amount='5', amount='5', paymentTransactionExternalKey='829e26a5-1592-440e-9fe7-4d861adaa718', pspReference='8514725091166343', PaymentModificationResponse{adyenCallErrorStatus=null, pspReference='8814725091172702', response='[capture-received]', additionalData={}}, duration=340, error=false

op='authorize', merchantAccount='MyCityDealDE', amount='10', amount='10', paymentTransactionExternalKey='8666c104-6738-49e2-8727-d727516717d2', PurchaseResult{result='Authorised', authCode='8047', pspReference='7914725091176134', reason='null', resultCode='Authorised', reference='null', paymentTransactionExternalKey='8666c104-6738-49e2-8727-d727516717d2', adyenCallErrorStatus=null, additionalData={authCode='8047', refusalReasonRaw='AUTHORISED', avsResult='5 No AVS data provided', avsResultRaw='5', acquirerReference='7F54NDV25OG', cvcResult='1 Matches', cvcResultRaw='1', acquirerCode='TestPmmAcquirer'}}, duration=525, error=false
op='cancel', merchantAccount='MyCityDealDE', paymentTransactionExternalKey='c5afe094-40bb-462e-a026-caecd1285300', pspReference='7914725091176134', PaymentModificationResponse{adyenCallErrorStatus=null, pspReference='8814725091182719', response='[cancel-received]', additionalData={}}, duration=183, error=false

op='authorize', merchantAccount='MyCityDealDE', amount='10', amount='10', paymentTransactionExternalKey='43854743-a134-47b3-984c-e8003622c298', PurchaseResult{result='Authorised', authCode='24385', pspReference='7914725091186141', reason='null', resultCode='Authorised', reference='null', paymentTransactionExternalKey='43854743-a134-47b3-984c-e8003622c298', adyenCallErrorStatus=null, additionalData={authCode='24385', refusalReasonRaw='AUTHORISED', avsResult='5 No AVS data provided', avsResultRaw='5', acquirerReference='7CA6B5NE2JH', cvcResult='1 Matches', cvcResultRaw='1', acquirerCode='TestPmmAcquirer'}}, duration=411, error=false
op='capture', merchantAccount='MyCityDealDE', amount='10', amount='10', paymentTransactionExternalKey='5f2dd36b-3c6d-4b59-ae03-05f138b7d0bc', pspReference='7914725091186141', PaymentModificationResponse{adyenCallErrorStatus=null, pspReference='8514725091186382', response='[capture-received]', additionalData={}}, duration=386, error=false
op='refund', merchantAccount='MyCityDealDE', amount='10', amount='10', paymentTransactionExternalKey='46552eef-55a0-4ebb-bd9e-e074fd34045b', pspReference='8514725091186382', PaymentModificationResponse{adyenCallErrorStatus=null, pspReference='8814725091192726', response='[refund-received]', additionalData={}}, duration=202, error=false
```

And here are some log entries for notification handling:
```bash
op='notificationHandling' eventCode='AUTHORISATION', pspReference='4823660019473428', originalReference='null', success='true', reason='111647:7629:5/2014', merchantReference='325147059', merchantAccount='TestMerchant', duration=4, error=false
op='notificationHandling' eventCode='NOTIFICATION_OF_CHARGEBACK', pspReference='4679660098811789', originalReference='1805647532335033', success='true', reason='2013.04.12/08.09.20/821', merchantReference='1-6GNEZHGTGT', merchantAccount='TestMerchant', duration=0, error=false
op='notificationHandling' eventCode='NOTIFICATION_OF_CHARGEBACK', pspReference='4894660098821804', originalReference='1329636226549589', success='true', reason='2013.04.12/08.18.50/821', merchantReference='1-AYA5VKUZDP', merchantAccount='TestMerchant', duration=0, error=false
op='notificationHandling' eventCode='NOTIFICATION_OF_CHARGEBACK', pspReference='5299660098821820', originalReference='4328641331002023', success='true', reason='2013.04.12/08.26.54/672', merchantReference='1-J6X6UXKG4A', merchantAccount='TestMerchant', duration=0, error=false
op='notificationHandling' eventCode='CHARGEBACK_REVERSED', pspReference='4815580881292917', originalReference='null', success='true', reason='Chargeback reversed', merchantReference='1-AWSSVULBWX', merchantAccount='TestMerchant', duration=0, error=false
```

### CODE REVIEW

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java [hunk 1]
```diff
@@ -20,33 +20,130 @@
...
+        if (paymentData == null ) {
+            return;
         }
+
+        if (paymentData.getAmount() != null) {
+            buffer.append(", amount='")
+                  .append(paymentData.getAmount())
+                  .append("'");
+        }
+        if (paymentData.getAmount() != null) {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: Should it be `getCurrency` instead?

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java [hunk 2]
```diff
@@ -20,33 +20,130 @@
...
+                  .append("'");
+        }
+    }
+
+    private void appendUserData(final StringBuilder buffer, final UserData userData) {
         if (userData != null && userData.getShopperReference() != null) {
-            stringBuilder.append(", customerId='")
-                         .append(userData.getShopperReference())
-                         .append("'");
+            buffer.append(", customerId='")
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: Maybe we should rename it to `shopperReference` in the log to avoid any confusion (unsure if we refer to `customerId` anywhere else)?

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java [hunk 3]
```diff
@@ -20,33 +20,130 @@
...
+        logger.info(logBuffer.toString());
+    }
+
+    protected void logTransactionError(final String transactionType, final String pspReference, String merchantAccount, final PaymentData paymentData, final AdyenCallResult<?> adyenCall) {
+        final StringBuilder logBuffer = new StringBuilder();
+        appendTransactionType(logBuffer, transactionType);
+        appendMerchantAccount(logBuffer, merchantAccount);
+        appendPspReference(logBuffer, pspReference);
+        appendPaymentData(logBuffer, paymentData);
+        appendPspReference(logBuffer, pspReference);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: `appendPspReference` is twice.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java [hunk 4]
```diff
@@ -20,33 +20,130 @@
...
+    }
+
+    protected void logTransaction(final String transactionType, final String pspReference, String merchantAccount, final PaymentData paymentData, final PaymentModificationResponse response, final AdyenCallResult<?> adyenCall) {
+        final StringBuilder logBuffer = new StringBuilder();
+        appendTransactionType(logBuffer, transactionType);
+        appendMerchantAccount(logBuffer, merchantAccount);
+        appendPaymentData(logBuffer, paymentData);
+        appendPspReference(logBuffer, pspReference);
+        appendModificationResponse(logBuffer, response);
+        appendDuration(logBuffer, adyenCall.getDuration());
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: Should we log the duration in the error lines too?

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java [hunk 5]
```diff
@@ -20,33 +20,130 @@
...
+
+        logger.info(logBuffer.toString());
+    }
+
+    protected void logTransactionError(final String transactionType, final String pspReference, String merchantAccount, final PaymentData paymentData, final AdyenCallResult<?> adyenCall) {
+        final StringBuilder logBuffer = new StringBuilder();
+        appendTransactionType(logBuffer, transactionType);
+        appendMerchantAccount(logBuffer, merchantAccount);
+        appendPspReference(logBuffer, pspReference);
+        appendPaymentData(logBuffer, paymentData);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: Should be before `appendPspReference` for consistency.

File: src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java [hunk 6]
```diff
@@ -64,63 +64,59 @@ public String handleNotifications(final String input) {
...
 
     private void handleNotification(final NotificationRequestItem item) {
-        logger.info("Handling notification: eventCode='{}', pspReference='{}', originalReference='{}', success='{}', reason='{}', merchantReference='{}', merchantAccount='{}'",
+        Exception error = null;
+        final long startTime = System.currentTimeMillis();
+        long duration = 0L;
+        try {
+            final AdyenNotificationHandler adyenNotificationHandler = getAdyenNotificationHandler(item);
+            if (adyenNotificationHandler == null) {
+                logger.warn("No handler available - ignoring");
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java]: Maybe print the `item` for debugging?

File: src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java [hunk 7]
```diff
@@ -64,63 +64,59 @@ public String handleNotifications(final String input) {
...
+            }
+
+            adyenNotificationHandler.handleNotification(item);
+            duration = System.currentTimeMillis() - startTime;
+        } catch (final Exception e) {
+            duration = System.currentTimeMillis() - startTime;
+            error = e;
+        } finally {
+            final String message = String.format(
+                    "op='notificationHandling' eventCode='%s', pspReference='%s', originalReference='%s', success='%s', reason='%s', merchantReference='%s', merchantAccount='%s', duration=%d, error=%s",
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java]: Missing `,` before `eventCode`.

File: src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java [hunk 8]
```diff
@@ -64,63 +64,59 @@ public String handleNotifications(final String input) {
...
-            logger.warn("No handler available - ignoring");
-            return;
+                    item.getMerchantAccountCode(),
+                    duration,
+                    error != null);
+
+            if (error == null) {
+                logger.info(message);
+            } else {
+                logger.error(message, error);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java]: Nit: `warn` maybe?

File: src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java [hunk 9]
```diff
@@ -144,21 +140,30 @@ private SendNotification parse(final InputStream inputStream) throws ParserConfi
...
-
-        return outputStream;
+    private ByteArrayOutputStream createSendNotificationResponse(final String value) {
+        try {
+            final SendNotificationResponse response = new SendNotificationResponse();
+            response.setNotificationResponse(value);
+
+            // Use SOAPMessage to add the envelope
+            final Document document;
+                document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java]: Extra line?

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java [hunk 10]
```diff
@@ -42,8 +42,6 @@
 
 public class AdyenPaymentServiceProviderHostedPaymentPagePort extends BaseAdyenPaymentServiceProviderPort implements Closeable {
 
-    private static final Logger logger = LoggerFactory.getLogger(AdyenPaymentServiceProviderHostedPaymentPagePort.class);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java]: Why this change?

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderPort.java [hunk 11]
```diff
@@ -39,20 +39,19 @@
 import org.killbill.billing.plugin.adyen.client.model.SplitSettlementData;
 import org.killbill.billing.plugin.adyen.client.model.UserData;
 import org.killbill.billing.plugin.adyen.client.payment.builder.AdyenRequestFactory;
-import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class AdyenPaymentServiceProviderPort extends BaseAdyenPaymentServiceProviderPort implements Closeable {
 
-    private static final Logger logger = LoggerFactory.getLogger(AdyenPaymentServiceProviderPort.class);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderPort.java]: Why this change?

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java (hunk 4 above)
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: Oh, I think I understand - it's part of `AdyenCallResult` in these cases?

**Comment by pierre** [comment]
Few remarks but 👍 overall!

File: src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java (hunk 8 above)
Comment (andrenpaes) [inline src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java]: I was thinking of using the error level for all error scenarios (which reminds me I need to fix the `logTransactionError` method), but I don't really have a strong opinion about it.

Is KB core using warn for exception handling?

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java (hunk 4 above)
Comment (andrenpaes) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: Yeah, that's correct. I kept using the `toString` for errors on that. But we could expand the result call for failures in the class if you prefer. Maybe it's more clear?

File: src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java (hunk 6 above)
Comment (andrenpaes) [inline src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java]: Not sure if it's worth it. Current implementation of `canHandleNotification` always returns true, so this should never happen.

To be honest, I'm not even sure we need whole "factoryish" method thing, but that's a different refactor.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java (hunk 1 above)
Comment (andrenpaes) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: Actually, this was a copy/paste duplication error, but sure adding the currency is a good idea 😄

Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: 👍 

In general, as a rule of thumb, we always try to handle amount and currency together in Kill Bill and plugins (in method signatures, database tables, logging, etc.) because an amount by itself isn't meaningful.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java (hunk 4 above)
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: > But we could expand the result call for failures in the class if you prefer. Maybe it's more clear?

Not strong opinion but I do think indeed that it would be clearer for the reader.

File: src/main/java/org/killbill/billing/plugin/adyen/client/jaxws/LoggingOutInterceptor.java [hunk 12]
```diff
@@ -34,7 +34,8 @@ public LoggingOutInterceptor(final Iterable<Obfuscator> obfuscators) {
     protected String transform(final String originalLogString) {
         String result = originalLogString.replaceAll("cvc>[0-9]+</", "cvc>***</")
                                          .replaceAll("number>[0-9]+</", "number>***</")
-                                         .replaceAll("bankAccountNumber>[0-9]+</", "bankAccountNumber>***</");
+                                         .replaceAll("bankAccountNumber>[0-9]+</", "bankAccountNumber>***</")
+                                         .replaceAll("Authorization=\\[[^\\]]*\\]", "Authorization=[****]");
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/jaxws/LoggingOutInterceptor.java]: 👍 

Nit: might make sense to enhance the [PatternObfuscator](https://github.com/killbill/killbill/blob/73df2253ad000c3596193983928cc3a449184770/profiles/killbill/src/main/java/org/killbill/billing/server/log/obfuscators/PatternObfuscator.java#L27) (which we could make it configurable more easily...).

Comment (andrenpaes) [inline src/main/java/org/killbill/billing/plugin/adyen/client/jaxws/LoggingOutInterceptor.java]: Good idea.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java (hunk 4 above)
Comment (andrenpaes) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/BaseAdyenPaymentServiceProviderPort.java]: I think moving everything to the base class would too much of a change. For now, I'll remove the duration from being printed in the `toString` method, and explicitly add it in the log* methods.

**Comment by sbrossie** [comment]
👍
